File: EJERCICIO_BANCO.py
class Banco:
    def __init__(self):
        self.employees = []

    def add_employee(self, employee):
        self.employees.append(employee)

    def get_employees(self):
        return self.employees

    def show_all_employees(self):
        for employee in self.employees:
            print("Empleado:", employee.get_name())
            print("Información:", employee.get_data())
            print("Cuentas Bancarias:")
            for account in employee.get_accounts():
                print("Numero de cuenta:", account.get_account_number())
                print("Tipo de cuenta:", account.get_account_type())
                print("Saldo:", account.get_balance())
            print("-----------------------")

    def get_employee_by_name(self, employee_name):
        for employee in self.employees:
            if employee.get_name() == employee_name:
                return employee
        return None

    def handle_account_operations(self, employee):
        employee.handle_account_operations()


class BankAccount:
    next_account_number = 1

    def __init__(self, account_type):
        self.account_number = BankAccount.next_account_number
        BankAccount.next_account_number += 1
        self.balance = 0.0
        self.account_type = account_type

    def deposit(self, amount):
        self.balance += amount
        print(f"Se han agregado ${amount:.2f} a la cuenta.")

    def withdraw(self, amount):
        if self.balance - amount >= 0:
            self.balance -= amount
            print(f"Se han retirado ${amount:.2f} de la cuenta.")
        else:
            print("Saldo insuficiente para retirar.")

    def get_balance(self):
        return self.balance

    def get_account_type(self):
        return self.account_type

    def get_account_number(self):
        return self.account_number


class Employee:
    def __init__(self, name, age, salary):
        self.name = name
        self.age = age
        self.salary = salary
        self.accounts = []
        self.determine_account_type()

    def determine_account_type(self):
        if self.salary <= 50000:
            self.add_account_type('A')
        elif self.salary <= 100000:
            self.add_account_type('B')
        else:
            self.add_account_type('C')

    def add_account_type(self, account_type):
        account = BankAccount(account_type)
        self.accounts.append(account)

    def get_accounts(self):
        return self.accounts

    def get_name(self):
        return self.name

    def get_age(self):
        return self.age

    def get_salary(self):
        return self.salary

    def get_data(self):
        return f"Nombre: {self.get_name()}, Edad: {self.get_age()}, Salario: {self.get_salary():.2f}"

    def handle_account_operations(self):
        if self.accounts:
            print("Seleccione una cuenta para operar:")
            for i, account in enumerate(self.accounts, 1):
                print(f"{i}. Numero de cuenta: {account.get_account_number()}, Tipo de cuenta: {account.get_account_type()}, Saldo: {account.get_balance()}")
            
            account_index = int(input("Ingrese el numero 1 para continuar con el proceso: "))
            if 1 <= account_index <= len(self.accounts):
                selected_account = self.accounts[account_index - 1]
                print("Operaciones disponibles:")
                print("1. Agregar dinero")
                print("2. Retirar dinero")
                option = int(input("Seleccione una opcion: "))
                if option == 1:
                    deposit_amount = float(input("Ingrese la cantidad a agregar: "))
                    selected_account.deposit(deposit_amount)
                elif option == 2:
                    withdraw_amount = float(input("Ingrese la cantidad a retirar: "))
                    selected_account.withdraw(withdraw_amount)
                else:
                    print("Opcion no valida.")
            else:
                print("Numero de cuenta invalido.")
        else:
            print("El empleado no tiene cuentas bancarias.")


if __name__ == "__main__":
    banco = Banco()

    while True:
        print("\nBIENVENIDO AL BANCO DE RYAN GOSLING")
        print("1. Agregar Nuevo Empleado")
        print("2. Mostrar todos los Empleados y sus Cuentas")
        print("3. Mostrar Informacion de un Empleado y Operaciones")
        print("4. Salir")
        option = int(input("Ingrese la opcion deseada: "))

        if option == 1:
            name = input("Ingrese el nombre del empleado: ")
            age = int(input("Ingrese la edad del empleado: "))
            salary = float(input("Ingrese el salario del empleado: "))

            new_employee = Employee(name, age, salary)
            banco.add_employee(new_employee)

            print("Empleado agregado con exito.")
        elif option == 2:
            print("Todos los Empleados y sus Cuentas:")
            banco.show_all_employees()
        elif option == 3:
            search_name = input("Ingrese el nombre del empleado a buscar: ")
            employee = banco.get_employee_by_name(search_name)
            if employee is not None:
                banco.handle_account_operations(employee)
            else:
                print("Empleado no encontrado.")
        elif option == 4:
            print("Hasta luego!")
            break
        else:
            print("Opción no valida. Intente nuevamente.")

File: EJERCICIO_BANCO.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Main.java to edit this template
 */

/*Realizar dos metodos para poder agrerar o retirar dinero de la cuenta bancaria. Considerar la siguiente:
La cuenta bancaria debe de ser alguno de estos tipos: A,B o C y debe de incluirse al momento de dar de alta
el momento de registrar un usuario con una cuenta. Si se ingresa un valor diferente a esos 3, advertirle al
usuario y no permitir darlo de alta.

-Si el numero de cuenta es de tipo A la cuenta puede tener hasta $50,000
-Si el numero de cuenta es de tipo B la cuenta puede tener hasta $100,000
-Si el numero de cuenta es de tipo C la cuenta puede tener saldo ilimitado

-Al momento de retirar dinero de la cuenta el saldo minimo para la cuenta de tipo A es de $1,000
-Al momento de retirar dinero de la cuenta el saldo minimo para la cuenta de tipo B es de $5,000
-Al momento de retirar dinero de la cuenta el saldo minimo para la cuenta de tipo C es de $10,000
 */


package ejercicio_banco;

import java.util.Scanner;

public class EJERCICIO_BANCO {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Banco banco = new Banco();

        while (true) {
            System.out.println("\nBIENVENIDO AL BANCO DE RYAN GOSLING");
            System.out.println("1. Agregar Nuevo Empleado");
            System.out.println("2. Mostrar todos los Empleados y sus Cuentas");
            System.out.println("3. Mostrar Informacion de un Empleado y Operaciones");
            System.out.println("4. Salir");
            System.out.print("Ingrese la opcion deseada: ");
            int option = scanner.nextInt();

            switch (option) {
                case 1:
                    System.out.print("Ingrese el nombre del empleado: ");
                    scanner.nextLine(); 
                    String name = scanner.nextLine();
                    System.out.print("Ingrese la edad del empleado: ");
                    int age = scanner.nextInt();
                    System.out.print("Ingrese el salario del empleado: ");
                    double salary = scanner.nextDouble();

                    Employee newEmployee = new Employee(name, age, salary);
                    banco.addEmployee(newEmployee);

                    System.out.println("Empleado agregado con exito.");
                    break;
                case 2:
                    System.out.println("Todos los Empleados y sus Cuentas:");
                    banco.showAllEmployees();
                    break;
                case 3:
                    System.out.print("Ingrese el nombre del empleado a buscar: ");
                    scanner.nextLine(); // Consumir la nueva línea pendiente
                    String searchName = scanner.nextLine();
                    Employee employee = banco.getEmployeeByName(searchName);
                    if (employee != null) {
                        banco.handleAccountOperations(employee, scanner);
                    } else {
                        System.out.println("Empleado no encontrado.");
                    }
                    break;
                case 4:
                    System.out.println("Hasta luego!");
                    scanner.close();
                    System.exit(0);
                    break;
                default:
                    System.out.println("Opcion no valida. Intente nuevamente.");
            }
        }
    }
}
